Step2 : 수강신청(도메인 모델)

Mo-bile opened the pull request:
안녕하세요? 리뷰요청드립니다

프로그래밍 요구사항대로 하면서 점점 실무와 연결되는 느낌이 있었습니다! 좋은 과정 감사드립니다
요구조건이 몇 부분 모호해서 그런지(실무도 사실 그렇죠 😂) 개발하면서 생각보다 많은 시간이 소요되었네요
일단은 한번이라도 더 리뷰 요청드리고 다시 리팩터링 하는것이 바람직하다고 생각들어서 요청드립니다

다만 개발을 하면서 가장 고민되는 부분이 있었습니다.

1. 무료인 경우와 유료인 경우를 구분하는 것인데요
특히 무료인 경우 수강료와 정원을 0 으로 두어야할지 고민이 되었습니다.
0 으로 두기에는 객체에 대한 도메인을 오염시킨다는 생각이 들었습니다.

그래서 null 이라는 것이 프로그램적으로는 위험할 수 있지만, 의미 그대로 존재하지 않는다 생각해서 null 로 이용했습니다
혹시 이렇게 접근하는게 괜찮을까요?

2. 무료 유료를 구분하면서 `ProvidePolicy` 를 초기화 하는데 `new ProvidePolicy()` 빈 생성자인 경우를 무료로 정의했습니다.
유료인 경우는 `new ProvidePolicy(x, y)` 인자가 있는 경우로 정의 했습니다.
이렇게 접근하는 것이 개발하는 과정에서 계속 불편했는데요 이렇게 접근이 괜찮을까요?

아니면 고민하다가 전략패턴으로 도전해볼까 고민했는데, 오버엔지니어링이라 생각들어서 일단 시도 하지 않았습니다.

3. 상위 객체 메서드부터 여러 단계의 하위 객체의 메서드까지 `파라미터 터널링` 이 몇부분 보이는데요 이부분도 걱정이됩니다.

감사합니다 👍

Commit:
refactor : 접근제어자 변경, 컨벤션 준수

Commit:
doc : Question의 필드의 객체 분리 필요성 관련 의견 메모

Commit:
fix : 컴파일 에러 해결

Commit:
feat : 강의 시작일 종료일 관련 객체 구현 및 유효성 검증

Commit:
feat : 강의 커버 이미지 객체 구현 및 정규식 검증

Commit:
feat : 강의 상태 기능 개발 및 테스트

Commit:
feat : 무료, 유료 강의에 따라 강의 신청 및 정원 지불 관리 개발

Commit:
refactor : enum 클래스 공통된 enumerate 패키지로 이동

Commit:
refactor : 정책관련 객체 ProvidePolicy 를 추가하고, Provide 에 일부 필드 변경

Commit:
feat : session 및 관련 객체 개발 완료, 수강신청 기능 개발 완료

Commit:
refactor : apply 관련 메서드 반환 값 변경 및 관련 테스트 코드 변경

Commit:
fix : 강의 모입 중 상태에 신청 가능하게 수정

Commit:
test : 깨진 테스트 고치기

Commit:
refactor
1. amount, pay 관련 자료형 int 에서 Long으로 변경
2. Session 에 session 찾는 메서드 추가

Commit:
feat : Course 에 예외전파 및 수강신청 추가

Commit:
refactor : 수강신청 수강생 수가 아니라 신청 자 리스트를 받게끔 변경

Commit:
feat : 결제모듈 붙이고, 항상 결제된것으로 구현

Commit:
refactor : EnrolledUsers, Sessions 일급컬렉션으로 변환

Commit:
refactor : 정원초과 로직 EnrolledUsers 로 이관

Commit:
docs : 개발 과정 메모

javajigi commented on src/main/java/nextstep/courses/domain/Base.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.time.LocalDateTime;
+
+public class Base {
```
```suggestion
public abstract class Base {
```
추상 클래스로 구현해 직접 생성할 수 없도록 구현

Mo-bile replied:
추상 클래스로 변경하여 컴파일러가 직접 생성할 수 없게끔 보장하게 변경하고 오직 직접 생성용으로 한다
결국 공통상태 관리하는 역할로 둔다.

javajigi commented on src/main/java/nextstep/courses/domain/Base.java:
```diff
@@ -0,0 +1,18 @@
+
+public class Base {
+    
+    protected LocalDateTime createdDate;
```
```suggestion
    private final LocalDateTime createdDate;
```
부모 객체 필드 또한 private으로 구현
값을 접근할 필요가 있으면 protected getter 메서드를 통해 접근

javajigi commented on src/main/java/nextstep/courses/domain/Course.java:
```diff
@@ -37,17 +41,35 @@ public Long getCreatorId() {
+    
+    public void apply(long userId, long sessionId, Payment payment) throws CanNotJoinException {
+        Session session = sessions.findToApplySession(sessionId);
+        if(payment == null) {
```
payment 값에 따라 결정하기보다 Session의 ProvideType에 따라 결정되도록 해야하지 않을까?

javajigi commented on src/main/java/nextstep/courses/domain/session/CoverImage.java:
```diff
@@ -0,0 +1,48 @@
+import nextstep.courses.CanNotCreateException;
+import nextstep.courses.enumerate.CoverImageType;
+
+public class CoverImage {
```
이 객체 또한 필드 수가 많다.
관련있는 값을 묶어 필드 수를 줄일 수 있지 않을까?

javajigi commented on src/main/java/nextstep/courses/domain/session/Duration.java:
```diff
@@ -0,0 +1,40 @@
+import java.time.LocalDate;
+import nextstep.courses.CanNotCreateException;
+
+public class Duration {
```
👍

javajigi commented on src/main/java/nextstep/courses/domain/enrollment/EnrolledUsers.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+import nextstep.courses.CanNotJoinException;
+
+public class EnrolledUsers {
```
👍

javajigi commented on src/main/java/nextstep/courses/domain/Provide.java:
```diff
@@ -0,0 +1,46 @@
+    }
+    
+    public void applyPaid(EnrolledUsers enrolledUsers, Payment payment) throws CanNotJoinException {
+        if(this.type == ProvideType.FREE) {
```
```suggestion
        if(this.type.isFree()) {
```
enum을 class와 같이 메시지를 보내는 방식으로 구현 가능

javajigi commented on src/main/java/nextstep/courses/domain/ProvidePolicy.java:
```diff
@@ -0,0 +1,68 @@
+import nextstep.courses.enumerate.ProvideType;
+import nextstep.payments.domain.Payment;
+
+public class ProvidePolicy {
```
수강 신청 가능 여부를 판단할 수 있는 모든 값을 가지도록 구현하면 어떨까?
예를 들어 SessionStatus, 현재 수강신청한 수강생 목록도 수강 신청 여부를 판단하는데 필요한 값으로 판단할 수 있지 않을까?

javajigi commented on src/main/java/nextstep/courses/domain/ProvidePolicy.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+    
+    public void isAvailableEnroll(EnrolledUsers enrolledUsers) throws CanNotJoinException {
```
```suggestion
    public void enroll(EnrolledUsers enrolledUsers) throws CanNotJoinException {
```
보통 is로 시작하는 메서드는 반환값이 boolean을 것을 기대함
AI를 통해 더 적절한 이름을 추천 받아보면 어떨까?

javajigi commented on src/test/java/nextstep/courses/domain/CourseTest.java:
```diff
@@ -0,0 +1,86 @@
+    
+    static {
+        try {
+            freeSession = new Session(
```
Session 객체와 같이 인스턴스 변수가 많은 객체를 테스트하려면 객체를 생성하는데 어려움이 있다.
중복 코드 또한 많이 발생해 Session을 생성할 때 생성자의 인자가 변경되는 경우 변경할 부분이 많아진다.
https://www.arhohuttunen.com/test-data-builders/ 문서 참고해 Session과 같이 복잡한 객체의 테스트 데이터를 생성할 때 어떤 방법을 사용할 것인지 선택해 보면 좋겠다.
이번 기회에 내가 선호하는 방법을 적용해 보고 앞으로도 쭈욱 활용하는 방식이면 좋겠다.

javajigi reviewed (changes requested):
2단계는 역시나 테이블과 매핑하지 않아 작은 단위의 여러 객체로 분리 👍 
전체적인 객체 설계와 단위 테스트 구현 잘 했네요.
그럼에도 불구하고 한번 더 고민해 봤으면 하는 부분이 있어 피드백 남겨요.


Q1. 무료인 경우와 유료인 경우를 구분하는 것인데요
    특히 무료인 경우 수강료와 정원을 0 으로 두어야할지 고민이 되었습니다.
    0 으로 두기에는 객체에 대한 도메인을 오염시킨다는 생각이 들었습니다.

그래서 null 이라는 것이 프로그램적으로는 위험할 수 있지만, 의미 그대로 존재하지 않는다 생각해서 null 로 이용했습니다
혹시 이렇게 접근하는게 괜찮을까요?

A1. null로 구현하기보다 null을 추상화한 Null Object를 추가해 보는 것은 어떨까요?
AI에게 Null Object에 대해 질문해 보고 피드백을 받아 적용해 볼 것을 추천합니다.

Q2. 무료 유료를 구분하면서 ProvidePolicy 를 초기화 하는데 new ProvidePolicy() 빈 생성자인 경우를 무료로 정의했습니다.
    유료인 경우는 new ProvidePolicy(x, y) 인자가 있는 경우로 정의 했습니다.
    이렇게 접근하는 것이 개발하는 과정에서 계속 불편했는데요 이렇게 접근이 괜찮을까요?

아니면 고민하다가 전략패턴으로 도전해볼까 고민했는데, 오버엔지니어링이라 생각들어서 일단 시도 하지 않았습니다.

A2. 무료 유료 구분에 따른 구분은 ProvideType 값에 따라 결정되어야 하지 않을까요?
생성자 유무로 판단하는 것은 적절해 보이지 않은 것 같아요.

Q3. 상위 객체 메서드부터 여러 단계의 하위 객체의 메서드까지 파라미터 터널링 이 몇부분 보이는데요 이부분도 걱정이됩니다.

A3. 피드백에도 남겼는데요.
상위 객체의 필드도 private으로 구현하고 메서드를 통해 접근할 것을 추천합니다.

Commit:
refactor :
1. Base를 추상 클래스로 구현하여 상속전용으로만 하고 직접 생성은 자식이 못하게 막음 (도메인 객체처럼 사용할 우려 막음)
2. Base 의 필드를 private 로 변경해 직접 수정 못하게 막음

Commit:
refactor : CoverImage 의 많은 필드의 원시값을 포장하여 관련된 것 끼리 분리

Commit:
refactor
1. type 검증은 enum 내부에서 직접 하게 변경
2. boolean 타입이 아닌데 is 로 시작한 메서드명 변경

Commit:
refactor : apply 에 ProvideType 에 따라 분기되게끔 변경

Commit:
refactor
1. 객체의 역할과 책임을 명확히 인지하기 위해 객체명 변경
2. Session 객체 필드 일부를 EnrollmentPolicy로 이관

Commit:
refactor : 각각 변수를 SessionApply로 객체 분리

Commit:
refactor : 불필요한 static 등 제거

Commit:
refactor : null object + 전략 패턴 도입으로 무료 유료강의 정책에 대한 DI 방식 구현

Commit:
refactor : 디렉토리 구조 변경

Commit:
refactor : 테스트 픽스쳐를 '오브젝트 마더 패턴' + 빌더패턴(기본 값 숨기기, 빌더 덜쓰기, 팩토리 메서드를 이용해 도메인 강조, 비슷한 객체를 이용 시 코드 줄이기위한 but 사용) 사용하여 리팩터링

Commit:
docs : 리팩터링 기준 메모

Mo-bile commented:
안녕하세요?
알려 주신 부분을 숙지하고 이해하고 한땀한땀 적용하고 고민한다고 금번 피드백 요청은 시간이 소요된 것 같습니다.

말씀주신 Null object, 테스트 픽스쳐 줄이는 방법을 적용했습니다.
그러면서 롬복으로만 사용하던 빌더패턴을 구현해보면서 구체적으로 어떤 것인지 깨달았고, 팩토리 메서드의 중요성도 깨닫고, Null Object를 활용하면서 인터페이스 활용도 익힌것 같습니다. 감사합니다

질문 드립니다.
질문드리고 나서 혼자서 고민하다가 Builder 디렉토리 내 Free, Paid로 유사한 코드가 반복되는것을 통합하여 질문을 `해결`했습니다.
~1. Bulder 디렉토리 내 Free, Paid로 나누어져 유사한 코드가 반복적으로 나오는데 괜찮을까요?
복잡한 객체의 테스트 데이터를 생성을 위해 Builder 디렉토리 안에 있습니다.
그런데 유료, 무료인 경우에 따라서 유사한 파일이 두개로 갈라지는데요
유사한 코드가 Free, Paid로 반복적으로 나오는데 괜찮을까요?~

Commit:
refactor : Free, Paid로 나누어진 Builder를 하나로 합쳐서 간소화

Commit:
refactor : 놓친 필드 기본 설정 제거

Commit:
chore : 패키지 리스트럭쳐링

javajigi reviewed (approved):
피드백 반영 👍 
도메인 객체 설계부터 접근하다보니 정말 많은 도메인 객체가 추가되었네요.
이 객체들이 어떻게 db와 매핑될지 궁금하네요.
가능하면 도메인 객체를 유지한 상태로 db와 매핑하는 것에 도전해 보면 좋겠습니다.
